### ISSUE TITLE
time_stretch does not return the same signal when rate=1

### ISSUE THREAD

**Issue by zafarrafii** [librosa/librosa#1700]
**Describe the bug**
time_stretch does not return the same signal when rate=1.

**To Reproduce**
```
import librosa
import matplotlib.pyplot as plt

audio_signal, _ = librosa.load(librosa.ex('fishin'))
audio_signal_stretched = librosa.effects.time_stretch(audio_signal, rate=1)

plt.plot(audio_signal, label='original')
plt.plot(audio_signal_stretched, label='stretched')
plt.plot(audio_signal_stretched-audio_signal, label='difference')
plt.legend(loc='upper left')
plt.show()
```
![image](https://user-images.githubusercontent.com/22797461/233508393-60dd85aa-24d1-4339-b1f7-3fad2a601ff5.png)

**Expected behavior**
time_stretch should return the same signal when rate=1.

**Software versions**
```
Linux-5.4.0-146-generic-x86_64-with-debian-buster-sid
Python 3.7.13 (default, Mar 29 2022, 02:18:16) 
[GCC 7.5.0]
NumPy 1.18.5
SciPy 1.4.1
librosa 0.9.1
INSTALLED VERSIONS
------------------
python: 3.7.13 (default, Mar 29 2022, 02:18:16) 
[GCC 7.5.0]

librosa: 0.9.1

audioread: 2.1.9
numpy: 1.18.5
scipy: 1.4.1
sklearn: 1.0.2
joblib: 1.1.0
decorator: 5.1.1
soundfile: 0.9.0
resampy: 0.2.2
numba: 0.55.1

numpydoc: 1.4.0
sphinx: 5.0.2
sphinx_rtd_theme: None
sphinxcontrib.versioning: None
sphinx-gallery: None
pytest: 7.1.2
pytest-mpl: None
pytest-cov: None
matplotlib: 3.5.1
presets: None
```

Perhaps a simple fix would be to add an if statement which would return the input if rate=1, just like in the time_stretch function of the rubberband Python wrapper: https://pyrubberband.readthedocs.io/en/stable/generated/pyrubberband.pyrb.time_stretch.html Thanks for that wrapper by the way, it works great! I also noticed that some values go out of the [-1, 1] range, but this seems to happen when resampling the audio..

**Comment by bmcfee** [comment]
Thanks for catching that.  It would be an easy enough flag to add, either in the effects wrapper functions or in the phase_vocoder itself.  (Or maybe both?)

The one subtlety that we'd need to worry about is that naively returning the input directly if `rate=1` will lead to a reference copy, which is semantically just a bit different from the current behavior (which always returns a new array).



> I also noticed that some values go out of the [-1, 1] range, but this seems to happen when resampling the audio..

Yes, this can definitely happen, especially in pitch shifting which also adds a sinc interpolation step over the phase vocoding.  There's nothing that generally constrains the output of these processes to preserve min/max bounds from the input, but the inverse windowing in istft should be well behaved anyway.

**Comment by dpwe** [comment]
>
>
> Yes, this can definitely happen, especially in pitch shifting which also
> adds a sinc interpolation step over the phase vocoding. There's nothing
> that generally constrains the output of these processes to preserve min/max
> bounds from the input, but the inverse windowing in istft should be well
> behaved anyway.
>
Any kind of sharp frequency transition (i.e., an extreme anti-aliasing
filter, or the brick-wall edge exposed by upsampling) will result in
ringing, with up to ~9% overshoot of clipped peaks.
https://en.wikipedia.org/wiki/Gibbs_phenomenon

  DAn.

**Comment by bmcfee** [comment]
Note: this should be improved by properly accounting for expected phase advance between frames; will come back to this later on.

**Comment by bmcfee** [comment]
I did a bit more careful listening to the outputs of my proposed rewrite and the existing implementation, and I'm not yet happy with how the results sound.  Gonna try to find some more time to put into this before opening a PR.

**Comment by bmcfee** [comment]
Ok, finally had a minute to sit down and think about this a bit more carefully.

First, the existing vocoder implementation is a little obfuscated, and does some unnecessary steps in the phase differential estimation.  I spent a minute to streamline this a bit as follows, and it produces numerically equivalent results:

```python
from librosa import util
def phase_vocoder(
    D,
    *,
    rate,
    hop_length= None,
    n_fft = None,
):
    
    if n_fft is None:
        n_fft = 2 * (D.shape[-2] - 1)

    if hop_length is None:
        hop_length = int(n_fft // 4)

    time_steps = np.arange(0, D.shape[-1], rate, dtype=np.float64)

    # Create an empty output array
    shape = list(D.shape)
    shape[-1] = len(time_steps)
    d_stretch = np.zeros_like(D, shape=shape)

    # Phase accumulator; initialize to the first sample
    phase_acc = np.angle(D[..., 0])

    # Pad 0 columns to simplify boundary logic
    padding = [(0, 0) for _ in D.shape]
    padding[-1] = (0, 2)
    D = np.pad(D, padding, mode="constant")

    for t, step in enumerate(time_steps):
        columns = D[..., int(step) : int(step + 2)]

        # Weighting for linear magnitude interpolation
        alpha = np.mod(step, 1.0)
        mag = (1.0 - alpha) * np.abs(columns[..., 0]) + alpha * np.abs(columns[..., 1])

        # Store to output array
        d_stretch[..., t] = util.phasor(phase_acc, mag=mag)

        # Compute phase advance
        dphase = np.diff(np.unwrap(np.angle(columns), axis=-1), axis=-1)[..., 0]
        
        # Accumulate phase
        phase_acc += dphase

    return d_stretch
```

This is pretty inefficient, since we're repeatedly absing, angling, and unwrapping, when we could just do this all at once up front.

Replacing the loop with vectorized equivalents*, we get the following:

```python
def new_phase_vocoder(D, *, rate, hop_length=None, n_fft=None, kind='linear'):
    
    if n_fft is None:
        n_fft = 2 * (D.shape[-2] - 1)
        
    if hop_length is None:
        hop_length = int(n_fft // 4)

    time_in = np.arange(0, D.shape[-1], 1)

    # Build the interpolator for magnitudes
    fmag = scipy.interpolate.interp1d(time_in,
                                      np.abs(D),
                                      fill_value='extrapolate',
                                      kind=kind,
                                      copy=False,
                                      assume_sorted=True,
                                      axis=-1)

    # Compute the output time grid
    time_out = np.arange(0, D.shape[-1], rate)

    # Separate out the time grid into fractional and integer lookup parts
    lookup = np.ceil(time_out)
    
    # Compute the unwrapped phase differentials
    angles_in = np.angle(D)
    phase_diff = np.diff(np.unwrap(angles_in, axis=-1), axis=-1, prepend=0)

    # Compute angle by accumulating phase differential estimates    
    angles = np.cumsum(np.take(phase_diff, lookup.astype(int), axis=-1, mode='clip'), axis=-1)
    
    d_stretch = librosa.util.phasor(angles=angles, mag=fmag(time_out))
    
    return d_stretch
```
* it's not exactly equivalent.  It does produce identical phase estimates as the original implementation up to the final set of output frames, i.e., those where the phase propagation estimate would walk off the end of the input array.  I'm okay with sacrificing equivalence on those frames since it's not well defined anyway.

There's a bit of trickery going on where, where we ceil the output time so that phase differences come from the next frame, not the previous frame.  To make this work properly, we have to clip the call to `np.take`, which is essentially forcing a stationarity assumption on the final input frame.  (This, I believe, is reasonable.)

The above implementation does satisfy the `rate=1` property (input == output), and it's anywhere between 2-4x faster than the explicit loop implementation.  The quality is still... well, it's the same algorithm more or less.  But it's at least faster and a bit easier to parse.

I'll prep a PR on this when I have some more sparse time.

**Comment by bmcfee** [comment]
> time_stretch does not return the same signal when rate=1.

After chasing my tail on this a bit further, I noticed that the starting point of this issue is not actually true - at least when the input spectrogram is in double precision:

```python
In [7]: y, sr = librosa.load(librosa.ex('trumpet'))

In [8]: D = librosa.stft(y, dtype=np.complex128)

In [9]: D2 = librosa.phase_vocoder(D, rate=1)

In [10]: np.allclose(D, D2)
Out[10]: True
```

That we don't have exact equivalence in single precision mode (default) is just due to numerical precision and error accumulation.  While the accumulated error can be large in practice, I don't think there's actually a bug here at all.  It's possible that we could rewrite some bits to be a bit more numerically stable, but I'm not sure it's worth the effort.

Any objections to leaving things as is and closing this issue?

**Comment by bmcfee** [comment]
... I'll still push up a fix to handle the odd frame length case, since that's a legitimate bug.

**Comment by zafarrafii** [comment]
Thanks for taking care of this, Brian! Feel free to close the issue if you believe it's all good.

**Comment by bmcfee** [comment]
> ... I'll still push up a fix to handle the odd frame length case, since that's a legitimate bug.

This is fixed by #1814 - I'll close this issue out now.
